xonsh/aliases.py

```diff
@@ -54,6 +55,80 @@ def EXEC_ALIAS_RE():
     return re.compile(r"@\(|\$\(|!\(|\$\[|!\[|\&\&|\|\||\s+and\s+|\s+or\s+|[>|<]")
 
 
+class FuncAlias:
+    """Provides a callable alias for xonsh commands."""
+
+    attributes = ["__xonsh_threadable__", "__xonsh_capturable__"]
+
+    def __init__(self, name, func):
+        self.__name__ = self.name = name
+        self.func = func
+        for attr in self.attributes:
+            if (val := getattr(func, attr, None)) is not None:
+                self.__setattr__(attr, val)
+
+    def run(self, *args, **kwargs):
+        try:
+            return self.func(*args, **kwargs)
+        except Exception:
+            print_exception(f"Exception in {repr(self)}")
+
+    def __repr__(self):
+        r = {"name": self.name, "func": self.func}
+        for attr in self.attributes:
+            if (val := getattr(self, attr, None)) is not None:
+                r[attr] = val
+        return f"FuncAlias({repr(r)})"
+
+
+class FuncAlias0(FuncAlias):
+    def __call__(self):
+        return self.run()
+
+
+class FuncAlias1(FuncAlias):
+    def __call__(self, args=None):
+        return self.run(args)
+
+
+class FuncAlias2(FuncAlias):
+    def __call__(self, args=None, stdin=None):
+        return self.run(args, stdin)
+
+
+class FuncAlias3(FuncAlias):
+    def __call__(self, args=None, stdin=None, stdout=None):
+        return self.run(args, stdin, stdout)
+
+
+class FuncAlias4(FuncAlias):
+    def __call__(self, args=None, stdin=None, stdout=None, stderr=None):
+        return self.run(args, stdin, stdout, stderr)
+
+
+class FuncAlias5(FuncAlias):
+    def __call__(self, args=None, stdin=None, stdout=None, stderr=None, spec=None):
+        return self.run(args, stdin, stdout, stderr, spec)
+
+
+class FuncAlias6(FuncAlias):
+    def __call__(
+        self, args=None, stdin=None, stdout=None, stderr=None, spec=None, stack=None
+    ):
+        return self.run(args, stdin, stdout, stderr, spec, stack)
+
+
+FUNC_ALIAS_CLASSES = [
+    FuncAlias0,
+    FuncAlias1,
+    FuncAlias2,
+    FuncAlias3,
+    FuncAlias4,
+    FuncAlias5,
+    FuncAlias6,
+]
```

Comment:
these classes seems to duplicate the use case of PartialEvalAlias0 . Instead of creating the new class, if we are to handle exceptions from aliases specially, we can handle them in those PartialEvalAlias class or higher up in the stack.

Reply:
The number of arguments count will be the same right? can't we just catch the exception at top level?

Reply:
to show the original traceback, we can use `raise Err from OriginalException` syntax